### DISCUSSION TITLE
Writer vertical direction, candidates obsecure text.

### DISCUSSION THREAD

**Discussion by HubKing** [fcitx/fcitx5#1065] [General]
Maybe a fringe usage, but it seems that Fcitx5 is using a wrong place for vertical direction in LibreOffice Writer. It's displayed beneath the first letter, but it should be placed on the left of the text.

![image](https://github.com/fcitx/fcitx5/assets/10900489/cb19fa7d-2ccf-45d1-bfab-aa60d8b2e636)

If I use regular horizontal direction, the candidates are placed correctly.

![image](https://github.com/fcitx/fcitx5/assets/10900489/2f6cedd2-ec40-45d3-bece-90ccf7788704)

KDE Plasma 6, Wayland, LibreOffice Fresh.

**Comment by wengxt** [comment]
coordinate is send by application
please report to libreoffice

**Reply by HubKing** [reply]
But when I tested the same thing on Ubuntu 24 which uses Gnome + IBus on https://view.distrosea.com/ (online Linux testing), it didn't show the candidates on the text. Writer version was similar (Mine was 24.2.3.2 and the Ubuntu's was 24.2.2.2).

![image](https://github.com/fcitx/fcitx5/assets/10900489/3b48b411-7ffb-42f8-9d38-2f70ffa14735)

**Reply by wengxt** [reply]
For libreoffice, it's very tricky to answer. 
First of all, you're testing under different desktop ,and libreoffice has different toolkit implementation.

SAL_USE_VCLPLUGIN=gtk3 libreoffice

On Plasma 6 Kwin wayland
![图片](https://github.com/fcitx/fcitx5/assets/259684/b4a60631-abbe-4e51-8f72-53cd346648b6)

But I also do find a issue, when you start conversion with space, the client cursor is moved to the start of current segment, so that's why it seems to cover the text in some cases.

![图片](https://github.com/fcitx/fcitx5/assets/259684/33325166-1645-4d34-a168-a337b4630854)


But if you consider the horizontal case:
![图片](https://github.com/fcitx/fcitx5/assets/259684/f223891b-ffe0-431a-82ea-64adec5e4d55)
![图片](https://github.com/fcitx/fcitx5/assets/259684/01545428-01c7-4e82-9a44-6fd9bd36aabf)

This allows the popup to align with the current segment

So it's not really a wrong behavior, but I do hope libreoffice may try to help to optimize here.

**Reply by wengxt** [reply]
BTW, did you check the "Fix embedded preedit cursor at the beginning of the preedit" in fcitx's mozc config? You can try to uncheck it (default value is false now).

**Reply by HubKing** [reply]
> SAL_USE_VCLPLUGIN=gtk3 libreoffice

Things were as you said on my desktop, too. This made the candidate pop-up look like Ubuntu (i.e., not obscure the input text). but pressing the space made the pop-up obscure the input text. Well, I guess it won't get fixed due to low use usage (few Japanese use Linux, and even few of them would create vertical documents). LibreOffice already has tons of old bugs that have not been fixed, and they won't fix this for a few users. So I think I have to use it as it is.

 "Fix embedded preedit cursor at the beginning of the preedit"  was unchecked in my case.

**Reply by HubKing** [reply]
Instead of using an online Ubuntu, i installed XUbuntu in my virtual machine and tested it more precisely. It seems that LibreOffice Writer + XFCE (Gnome fork) + IBus + Mozc, vertical writing works perfectly. Please see the screen recording below (I normally use Kana input, but in this video, I had to use Romaji input, so there were lots of mistypes.):

[Screencast_20240526_224628.webm](https://github.com/fcitx/fcitx5/assets/10900489/bcef41fe-4788-4739-b715-785ce6f603d6)

**Comment by wengxt** [comment]
You didnt  try to show a long candidates list like my screenshot, can you
try that?

I'm pretty sure ibus and fcitx use the exact logic for set mozc cursor.

GitHub King ***@***.***> 于 2024年5月26日周日 06:51写道：

**Reply by HubKing** [reply]
I had used Xubuntu because I had the ISO downloaded before. But XFCE is using X11, and I suspected that it may not work well under Wayland. So, I downloaded the regular Ubuntu 24 which uses Gnome/Wayland, and tested it. It worked almost perfectly, except that for some reason, the "Enter" key did not work. See the screen recording below. I think I did the "long candidate" you mentioned. I wanted to test KDE Plasma 6 X11 + Fcitx5 if it fixes the problem, if the cause is Wayland, but I failed to make Fcitx5 work in the Plasma 6 X11 session.

[ubuntu.webm](https://github.com/fcitx/fcitx5/assets/10900489/95679dc7-bd2c-45f1-a54a-70bfd48b731c)

**Reply by HubKing** [reply]
Seems like Fcitx5 works the same way when LibreOffice is using GTK3. So, the bug is in the QT6 front-end of LibreOffice?

**Reply by wengxt** [reply]
> Seems like Fcitx5 works the same way when LibreOffice is using GTK3. So, the bug is in the QT6 front-end of LibreOffice?

Yes.

**Reply by wengxt** [reply]
No need to try with vertical, with normal horizontal may demonstrate the issue more easily
Gtk3
![gtk3](https://github.com/fcitx/fcitx5/assets/259684/6d779d41-4115-45db-bfea-d41223e02031)
Qt6
![qt6](https://github.com/fcitx/fcitx5/assets/259684/134bd879-d305-4f98-99b6-189500ce0d1e)

**Reply by wengxt** [reply]
I opened at bug at https://bugs.documentfoundation.org/show_bug.cgi?id=161286

**Reply by HubKing** [reply]
Thanks for the bug report. It is very much helpful both for me and for the developers at LibreOffice that someone who has deep knowledge on the inner workings of IME opened it instead of me, who would have just written "It doesn't work correctly under KDE Plasma".
